docs: Add REST API page and document OIDC provisioning

jamilbk opened the pull request:
- Fixes #6035 
- Clarifies how to provision users and groups when setting up the OIDC connector

thomaseizinger committed:
test(connlib): track pending connections to gateways (#6497)

Instead of tracking pending connections to resources, we need to model
pending connections to gateways. The offending test seed has a CIDR
resource that is a DNS server and the Internet resources, both routed
via the same gateway.

When sending concurrent DNS queries to those resources, we need to track
which _gateways_ we are connecting to as a result to figure out which
queries get lost. In particular, only the _first_ resource to trigger a
connection to a gateway will be authorized. Subsequent queries will be
completely lost and require another packet to authorize the connection.

---------

Signed-off-by: Thomas Eizinger <thomas@eizinger.io>
Co-authored-by: Not Applicable <ReactorScram@users.noreply.github.com>

AndrewDryga committed:
chore(portal): Add session data to fields truncated by logger (#6572)

jamilbk committed:
Add REST API page and document OIDC behavior

Patticatti committed:
fix(website): Fix overflowing logos animation (#6570)

dependabot[bot] committed:
build(deps): Bump com.google.firebase:firebase-bom from 33.1.2 to 33.2.0 in /kotlin/android (#6531)

Bumps com.google.firebase:firebase-bom from 33.1.2 to 33.2.0.


[![Dependabot compatibility
score](https://dependabot-badges.githubapp.com/badges/compatibility_score?dependency-name=com.google.firebase:firebase-bom&package-manager=gradle&previous-version=33.1.2&new-version=33.2.0)](https://docs.github.com/en/github/managing-security-vulnerabilities/about-dependabot-security-updates#about-compatibility-scores)

Dependabot will resolve any conflicts with this PR as long as you don't
alter it yourself. You can also trigger a rebase manually by commenting
`@dependabot rebase`.

[//]: # (dependabot-automerge-start)
[//]: # (dependabot-automerge-end)

---

<details>
<summary>Dependabot commands and options</summary>
<br />

You can trigger Dependabot actions by commenting on this PR:
- `@dependabot rebase` will rebase this PR
- `@dependabot recreate` will recreate this PR, overwriting any edits
that have been made to it
- `@dependabot merge` will merge this PR after your CI passes on it
- `@dependabot squash and merge` will squash and merge this PR after
your CI passes on it
- `@dependabot cancel merge` will cancel a previously requested merge
and block automerging
- `@dependabot reopen` will reopen this PR if it is closed
- `@dependabot close` will close this PR and stop Dependabot recreating
it. You can achieve the same result by closing it manually
- `@dependabot show <dependency name> ignore conditions` will show all
of the ignore conditions of the specified dependency
- `@dependabot ignore this major version` will close this PR and stop
Dependabot creating any more for this major version (unless you reopen
the PR or upgrade to it yourself)
- `@dependabot ignore this minor version` will close this PR and stop
Dependabot creating any more for this minor version (unless you reopen
the PR or upgrade to it yourself)
- `@dependabot ignore this dependency` will close this PR and stop
Dependabot creating any more for this dependency (unless you reopen the
PR or upgrade to it yourself)


</details>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
build(deps): Bump framer-motion from 11.3.28 to 11.3.31 in /website (#6533)

Bumps [framer-motion](https://github.com/framer/motion) from 11.3.28 to
11.3.31.
<details>
<summary>Changelog</summary>
<p><em>Sourced from <a
href="https://github.com/framer/motion/blob/main/CHANGELOG.md">framer-motion's
changelog</a>.</em></p>
<blockquote>
<h2>[11.3.31] 2024-08-29</h2>
<h3>Fixed</h3>
<ul>
<li>Detects conflicts between optimised animations and rendered
<code>style</code>.</li>
<li>Resumes optimised animations where possible once layout measurements
are complete.</li>
</ul>
<h2>[11.3.30] 2024-08-23</h2>
<h3>Fixed</h3>
<ul>
<li>Removed double <code>&quot;change&quot;</code> handler on
externally-provided <code>MotionValue</code>s.</li>
</ul>
<h2>[11.3.29] 2024-08-21</h2>
<h3>Fixed</h3>
<ul>
<li>Selective cancellation of optimised appear animations.</li>
</ul>
</blockquote>
</details>
<details>
<summary>Commits</summary>
<ul>
<li><a
href="https://github.com/framer/motion/commit/2947065346eeb1485417f94934b0cb793a0d5f81"><code>2947065</code></a>
v11.3.31</li>
<li><a
href="https://github.com/framer/motion/commit/aee687f636feca6c537f148f62a3450fd9c1f86a"><code>aee687f</code></a>
Updating changelog</li>
<li><a
href="https://github.com/framer/motion/commit/45a1d06d476f809bc4ef447a0c4bc1ddfd2f9d28"><code>45a1d06</code></a>
Resume optimised animations where possible (<a
href="https://redirect.github.com/framer/motion/issues/2774">#2774</a>)</li>
<li><a
href="https://github.com/framer/motion/commit/689b2476952c247486cd300bd9f5ddf929ead8fd"><code>689b247</code></a>
Bump webpack from 5.76.1 to 5.94.0 (<a
href="https://redirect.github.com/framer/motion/issues/2776">#2776</a>)</li>
<li><a
href="https://github.com/framer/motion/commit/d2484a88f1f553064d7450c72d65bbb9cc87b2b0"><code>d2484a8</code></a>
Ensure optimised appear animations are cancelled when styles differ (<a
href="https://redirect.github.com/framer/motion/issues/2772">#2772</a>)</li>
<li><a
href="https://github.com/framer/motion/commit/4c92d2fc9b0da9cd96cdc8b7995c6f6b4add5461"><code>4c92d2f</code></a>
Increasing bundlesizes</li>
<li><a
href="https://github.com/framer/motion/commit/9436cce05ab99a9d2f9b69b1e285e3df5f38a340"><code>9436cce</code></a>
updating version</li>
<li><a
href="https://github.com/framer/motion/commit/7f7e4305d45e0a1023c89dba0670bd0cefe17742"><code>7f7e430</code></a>
v11.3.30</li>
<li><a
href="https://github.com/framer/motion/commit/a0f0573082b42b44f784c86c2d7b681339af4e1d"><code>a0f0573</code></a>
Updating changelog</li>
<li><a
href="https://github.com/framer/motion/commit/828b8d9e52ee3db1693d5dad79a320f355e6e51f"><code>828b8d9</code></a>
Removing double update listenrs on externall provided motion values (<a
href="https://redirect.github.com/framer/motion/issues/2773">#2773</a>)</li>
<li>Additional commits viewable in <a
href="https://github.com/framer/motion/compare/v11.3.28...v11.3.31">compare
view</a></li>
</ul>
</details>
<br />


[![Dependabot compatibility
score](https://dependabot-badges.githubapp.com/badges/compatibility_score?dependency-name=framer-motion&package-manager=npm_and_yarn&previous-version=11.3.28&new-version=11.3.31)](https://docs.github.com/en/github/managing-security-vulnerabilities/about-dependabot-security-updates#about-compatibility-scores)

Dependabot will resolve any conflicts with this PR as long as you don't
alter it yourself. You can also trigger a rebase manually by commenting
`@dependabot rebase`.

[//]: # (dependabot-automerge-start)
[//]: # (dependabot-automerge-end)

---

<details>
<summary>Dependabot commands and options</summary>
<br />

You can trigger Dependabot actions by commenting on this PR:
- `@dependabot rebase` will rebase this PR
- `@dependabot recreate` will recreate this PR, overwriting any edits
that have been made to it
- `@dependabot merge` will merge this PR after your CI passes on it
- `@dependabot squash and merge` will squash and merge this PR after
your CI passes on it
- `@dependabot cancel merge` will cancel a previously requested merge
and block automerging
- `@dependabot reopen` will reopen this PR if it is closed
- `@dependabot close` will close this PR and stop Dependabot recreating
it. You can achieve the same result by closing it manually
- `@dependabot show <dependency name> ignore conditions` will show all
of the ignore conditions of the specified dependency
- `@dependabot ignore this major version` will close this PR and stop
Dependabot creating any more for this major version (unless you reopen
the PR or upgrade to it yourself)
- `@dependabot ignore this minor version` will close this PR and stop
Dependabot creating any more for this minor version (unless you reopen
the PR or upgrade to it yourself)
- `@dependabot ignore this dependency` will close this PR and stop
Dependabot creating any more for this dependency (unless you reopen the
PR or upgrade to it yourself)


</details>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

jamilbk committed:
Add to sidebar

jamilbk committed:
fix(website): Remove logo (#6585)

Requested.

Co-authored-by: Patti <139997703+Patticatti@users.noreply.github.com>

ReactorScram committed:
chore(rust/gui-client): remove unused version of `tauri-winrt-notification` (#6586)

We use a newer version directly, we don't need the one Tauri wraps.

Refs #6247

dependabot[bot] committed:
build(deps-dev): Bump @types/node from 22.0.2 to 22.5.2 in /rust/gui-client (#6553)

Bumps
[@types/node](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/node)
from 22.0.2 to 22.5.2.
<details>
<summary>Commits</summary>
<ul>
<li>See full diff in <a
href="https://github.com/DefinitelyTyped/DefinitelyTyped/commits/HEAD/types/node">compare
view</a></li>
</ul>
</details>
<br />


[![Dependabot compatibility
score](https://dependabot-badges.githubapp.com/badges/compatibility_score?dependency-name=@types/node&package-manager=npm_and_yarn&previous-version=22.0.2&new-version=22.5.2)](https://docs.github.com/en/github/managing-security-vulnerabilities/about-dependabot-security-updates#about-compatibility-scores)

Dependabot will resolve any conflicts with this PR as long as you don't
alter it yourself. You can also trigger a rebase manually by commenting
`@dependabot rebase`.

[//]: # (dependabot-automerge-start)
[//]: # (dependabot-automerge-end)

---

<details>
<summary>Dependabot commands and options</summary>
<br />

You can trigger Dependabot actions by commenting on this PR:
- `@dependabot rebase` will rebase this PR
- `@dependabot recreate` will recreate this PR, overwriting any edits
that have been made to it
- `@dependabot merge` will merge this PR after your CI passes on it
- `@dependabot squash and merge` will squash and merge this PR after
your CI passes on it
- `@dependabot cancel merge` will cancel a previously requested merge
and block automerging
- `@dependabot reopen` will reopen this PR if it is closed
- `@dependabot close` will close this PR and stop Dependabot recreating
it. You can achieve the same result by closing it manually
- `@dependabot show <dependency name> ignore conditions` will show all
of the ignore conditions of the specified dependency
- `@dependabot ignore this major version` will close this PR and stop
Dependabot creating any more for this major version (unless you reopen
the PR or upgrade to it yourself)
- `@dependabot ignore this minor version` will close this PR and stop
Dependabot creating any more for this minor version (unless you reopen
the PR or upgrade to it yourself)
- `@dependabot ignore this dependency` will close this PR and stop
Dependabot creating any more for this dependency (unless you reopen the
PR or upgrade to it yourself)


</details>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Patticatti committed:
fix(website): New subscription cards style (#6549)

Signed-off-by: Patti <139997703+Patticatti@users.noreply.github.com>

Patticatti committed:
fix(website): Change subscription plan text colour to neutral (#6588)

Signed-off-by: Patti <139997703+Patticatti@users.noreply.github.com>

AndrewDryga committed:
fix(portal): Reset the rate limit when user signed in (#6582)

jamilbk committed:
fix(portal): Use `app-ipv6.firezone.dev` for IPv6 app to prevent websocket / http from using different stacks (#6522)

Based on testing and research it does not appear that Chrome will
reliably choose a consistent protocol stack for loading the initial web
page as it does for connecting the WebSocket when connecting over VPN
tunnels. If one or the other stacks experiences a slight delay or packet
loss causing retransmission, or QUIC simply doesn't play nicely with the
MTU (in our case 1280), it may fall back to IPv4 (which has less
per-packet overhead) or even a TCP connection.

Unfortunately this violates an assumption we have in token validation
logic. Namely, that the remote_ip used to create the token (via sign in)
is the same one used to the connect the WebSocket. I can see where this
logic comes from in a security context, but thinking through the attack
vector(s) that would be able to leverage this violation has me left
wondering if this check is worth the breakage we currently face in
#6511.

- Scenario 1: MITM - attacker steals token somehow via MITM (would need
to somehow break TLS) - the attacker is already in our network path and
can rewrite the remote_ip already with his/her own.
- Scenario 2: Malicious browser plugin stealing session token. It will
be harder to spoof the remote IP in this case, but if this is a
possibility, the plugin could presumably directly control the tab where
the user is logged in.
- Scenario 3: IdP is compromised leading to malicious redirect before
arriving to Firezone - if this is the case, the user could likely login
in directly and create his/her own valid session token anyhow.

Perhaps I'm missing other scenarios, open to feedback. If we want to
ensure the token used by the websocket originated from the same browser
as it was minted from, perhaps we could generate a small random key,
save it in local storage, and send that in a header when connecting the
WebSocket. I think cookies handle that for us already though.

Fixes #6511

bmanifold commented on website/src/app/kb/reference/rest-api/readme.mdx:
```diff
@@ -0,0 +1,30 @@
+
+We publish an [OpenAPI spec](https://swagger.io/specification/) for the API that
+you can use to generate client libraries or explore the API interactively. You
+can find the spec at `https://api.firezone.dev/swaggerui`.
```
```suggestion
you can use to generate client libraries or explore the API interactively. You
can download the spec at [https://api.firezone.dev/openapi](https://api.firezone.dev/openapi) or view the spec with a Swagger UI web interface at [https://api.firezone.dev/swaggerui](https://api.firezone.dev/swaggerui).
```

bmanifold commented on website/src/app/kb/reference/rest-api/readme.mdx:
```diff
@@ -0,0 +1,30 @@
+`Settings -> API Clients` section of the admin portal. That will send a
+notification to the Firezone team to enable the API for your account.
+
+Once enabled, you can generate an API key and use it with any HTTP client.
```
Wondering if we should mention that they will need to create an API client first?

Also, just to keep terminology the same in the portal we use `API Token` rather than `API Key`

bmanifold commented on website/src/app/kb/authenticate/oidc/readme.mdx:
```diff
@@ -102,13 +98,25 @@ https://your-tenant.okta.com/.well-known/openid-configuration
+### Provisioning users and groups
+
+Users must be created in Firezone before they can authenticate with the
+universal OIDC connector. You can [create users](/kb/deploy/users) manually in
```
This could potentially be confusing to customers that are not on enterprise, because in the portal, you have the option to click on Google, Microsoft, Okta, Jump Cloud, or OIDC even if you're not on the enterprise plan.  All of those options will create a generic OIDC provider behind the scenes but from the user's perspective they may think they aren't using the generic OIDC provider.

Perhaps something like: 
'if you do not have directory sync enabled (enterprise only plans) then you will need to create users before.....'  or something along those lines.

jamilbk replied:
Hm, but we have dedicated guides for those connectors. This guide is just for the OIDC connector, i.e. when they click on the `OIDC` connector specifically.

bmanifold replied:
Ah, ok, I see.  I was thinking this was on a general 'Authentication' page.  Sorry about that.

jamilbk committed:
fix(website): Fix title space (#6589)

bmanifold commented on website/src/app/kb/authenticate/oidc/readme.mdx:
```diff
@@ -102,13 +98,25 @@ https://your-tenant.okta.com/.well-known/openid-configuration
-If the option is available, be sure to **enable PKCE** for the connector. This
-is a security feature that helps prevent certain types of attacks and is
-recommended for added security.
+The universal OIDC connector **requires** PKCE (Proof Key for Code Exchange) to
```
Similar to the comment below, many people may not realize they are using the universal OIDC provider.  We may want to say something like "All providers using OIDC **require** PKCE...."

jamilbk replied:
True, but we know that the other ones _do_ support OIDC, so PKCE really only apply to the universal connector I think

thomaseizinger committed:
refactor(connlib): only buffer 1 unsent packet if socket is busy (#6563)

Currently, we buffer UDP packets whenever the socket is busy and try to
flush them out at a later point. This requires allocations and is tricky
to get right.

In order to solve both of these problems, we refactor `snownet` to
return us an `EncryptedPacket` instead of a `Transmit`. An
`EncryptedPacket` is an indirection-abstraction that can be turned into
a `Transmit` given an `EncryptBuffer`. This combination of types allows
us to hold on to the `EncryptedPacket` (which does not contain any
references itself) in the `io` component whilst we are waiting for the
socket to be ready to send again.

This means we will immediately suspend the event loop in case the socket
is no longer ready for sending and resend the datagram in the
`EncryptBuffer` once we get re-polled.

jamilbk committed:
Update website/src/app/kb/reference/rest-api/readme.mdx

Co-authored-by: Brian Manifold <bmanifold@users.noreply.github.com>
Signed-off-by: Jamil <jamilbk@users.noreply.github.com>

jamilbk committed:
PR feedback

jamilbk committed:
Merge branch 'main' into docs/oidc

bmanifold reviewed (approved):
LGTM

jamilbk committed:
Fix typo
